Lib/test/test_unittest/testmock/testhelpers.py

```diff
@@ -1107,6 +1107,32 @@ class WithNonFields:
                 with self.assertRaisesRegex(AttributeError, msg):
                     mock.b
 
+    def test_dataclass_default_value_type_overrides_field_annotation(self):
+        # If field defines an actual default, we don't need to change
+        # the default type. Since this is how it used to work before #124176
+        @dataclass
+        class WithUnionAnnotation:
+            narrow_default: int | None = field(default=30)
+
+        for mock in [
+            create_autospec(WithUnionAnnotation, instance=True),
+            create_autospec(WithUnionAnnotation()),
+        ]:
+            with self.subTest(mock=mock):
+                self.assertIs(mock.narrow_default.__class__, int)
+
+    def test_dataclass_field_with_no_default_value(self):
+        @dataclass
+        class WithUnionAnnotation:
+            no_default: int | None
+
+        mock = create_autospec(WithUnionAnnotation, instance=True)
+        self.assertIs(mock.no_default.__class__, type(int | None))
```

Comment:
Same with these two, but this one is particularly confusing.
How can an object be both `None` and and `int` at the same time?

I know it's tangential to this PR, but I have to admit I'm not sure what the intention of `instance=True|False` is, the docs don't make it any clearer for me :-/

"You can use a class as the spec for an instance object by passing instance=True"

Does that mean:
- `create_autospec` will produce a mock instance based on treating `spec` as a class
-  you can produce a mock _class_ by passing in an instance as `spec` with `instance=True`
- something else?

@sobolevn / @ncoghlan - do either of you know?

Reply:
Some objects need real runtime resources to instantiate, so deriving an instance mock spec from an actual instance isn't practical. "instance = True" tells the module to do the best it can to mock an instance based on a class definition without needing to make a real instance.

3.14 is getting an enhancement to mock data class fields with no defaults based on their type annotations, but it has the same limitation in the absence of a default value as type checkers do: it can only narrow the field type for lazily populated fields down to a union, not to a concrete type.

This PR fixes a compatibility issue with the handling of data class fields that *do* have defaults (they get a mocked spec based on their default value instead of their declared union type).

I admit the handling of complex annotations isn't great (they are speced based on the annotation itself). Maybe we should treat the field annotation as an unconstrained mock (effectively Any) if the annotation doesn't correspond to a true runtime type?

Reply:
> annotations of the accepted types

Sorry, I didn't get this part. Can you please provide an example?

Reply:
@cjw296 this currently would be:

```python
>>> int | None
int | None
>>> type(int | None)
<class 'types.UnionType'>
```

This is a `typing` primitive. Docs: https://docs.python.org/3/library/types.html#types.UnionType

Reply:
The _instance_ attribute of the mock instance would be a type? That doesn't make any sense on the face of it, what am I missing?